pallets/collator-selection/src/lib.rs

```diff
@@ -500,29 +554,87 @@ pub mod pallet {
 			collators.extend(candidates.into_iter().collect::<Vec<_>>());
 			collators
 		}
-		/// Kicks out and candidates that did not produce a block in the kick threshold.
-		pub fn kick_stale_candidates(
+
+		/// Removes collators with unsatisfactory performance
+		/// Returns the removed AccountIds
+		pub fn evict_bad_collators(
 			candidates: Vec<CandidateInfo<T::AccountId, BalanceOf<T>>>,
 		) -> Vec<T::AccountId> {
-			let now = frame_system::Pallet::<T>::block_number();
-			let kick_threshold = T::KickThreshold::get();
-			candidates
-				.into_iter()
-				.filter_map(|c| {
-					let last_block = <LastAuthoredBlock<T>>::get(c.who.clone());
-					let since_last = now.saturating_sub(last_block);
-					if since_last < kick_threshold {
-						Some(c.who)
-					} else {
-						let outcome = Self::try_remove_candidate(&c.who);
-						if let Err(why) = outcome {
-							log::warn!("Failed to remove candidate {:?}", why);
-							debug_assert!(false, "failed to remove candidate {:?}", why);
-						}
-						None
+			use sp_runtime::PerThing;
+
+			// 0. Storage reads and precondition checks
+			if candidates.is_empty() {
+				return Vec::new(); // No candidates means we're running invulnerables only
+			}
+			let percentile_for_kick = Self::eviction_baseline();
+			if percentile_for_kick == Percent::zero() {
+				return Vec::new(); // Selecting 0-th percentile disables kicking. Upper bound check in fn build()
+			}
+			let underperformance_tolerated = Self::eviction_tolerance();
+			if underperformance_tolerated == Percent::one() {
+				return Vec::new(); // tolerating 100% underperformance disables kicking
+			}
+			let mut collator_perf_this_session =
+				<BlocksPerCollatorThisSession<T>>::iter().collect::<Vec<_>>();
+			if collator_perf_this_session.is_empty() {
+				return Vec::new(); // no validator performance recorded ( should not happen )
+			}
+
+			// 1. Ascending sort of collator performance list by number of produced blocks
+			collator_perf_this_session.sort_unstable_by_key(|k| k.1);
```

Comment:
Is unstable sort correct here? Is there any loss of fairness by doing this? It's going to depend on the ordering that comes out of `BlocksPerCollatorThisSession::iter`.

Reply:
We apply a tolerance (and even if that is 0%) and kick _below_ it due to this line:
https://github.com/Manta-Network/Manta/pull/358/files#diff-a0359a0add814d66aac74493f7e17aefdcee21d3de1ae592e97059082b6feedfR610

To my knowledge the only effect of unstable sorting is that a collators that produced equally many blocks as the one selected for the percentile will be randomly sorted above or below it, and are thus randomly added to the kick_candidates set and checked for underperformance.
As the checked nodes have the same created blocks count as the one at percentile, they will never be removed, so there is no impact to fairness in my view.

Reply:
Ok great, just checking.